### PR TITLE
[Supplier] Add `ServiceConfigs` to `SupplierStaking`

### PR DESCRIPTION
**Pull request by Olshansk**
## Summary

- Update all related components (CLI, Tx, Message, etc...) so suppliers can stake for a service

### Summary generated by Reviewpad on 31 Oct 23 22:36 UTC
This pull request includes changes to multiple files. Here is a summary of the changes:

1. `genesis.go`: The `Validate` function now checks the validity of stake values and delegatee addresses for the applications. It also validates the application service configurations using the `ValidateAppServiceConfigs` function instead of `AreValidAppServiceConfigs`.

2. `msg_server_stake_application.go`: Modifications in this file include an updated comment clarifying the validation of service configs.

3. `genesis.go` in the `x/supplier/types` directory: Changes in this file include the addition of an import statement for the `pocket/x/shared/helpers` package. It also adds validation for the stake value of suppliers, ensuring it is not nil, validation for the stake amount denom, checking if it is equal to "upokt", and validation for the application service configs of suppliers using the `servicehelpers.ValidateSupplierServiceConfigs` function.

4. `message_stake_supplier_test.go` in the `x/supplier/types` directory: Changes in this file include an added import statement for `sharedtypes`, a TODO_CLEANUP comment suggesting the simplification of a test, and test cases for validating the basic message of stake supplier and other related scenarios.

5. `genesis_test.go`: Changes in this file include an import statement for the `pocket/x/shared/types` package and the addition of two new elements to the `Genesis` test case, including the `ServiceConfigs` field with respective values.

6. `session.go`: Changes in this file include the addition of empty `Configs` slices to two structs, `sharedtypes.MainSession` and `sharedtypes.SupplierSession`.

7. `service_configs.go`: Changes in this file include the renaming of the function `AreValidAppServiceConfigs` to `ValidateAppServiceConfigs`, returning an error instead of a string and boolean value, changing the error messages to use `fmt.Errorf`, and the addition of the `ValidateSupplierServiceConfigs` function to validate supplier service configs.

8. `tx_stake_supplier_test.go`: Changes in this file include new test cases for stake supplier functionality and an update to an existing test case for stake supplier validation.

9. `message_stake_application.go`: Additions in this file include the `NewMsgStakeApplication` function, changes in the `ValidateBasic` function, and TODO comments related to technical debt.

10. `genesis_test.go`: Changes in this file include the addition of extensive tests related to the validity of the genesis state.

11. `service_configs.go`: Changes in this file include updates to the validation function, changes to error handling, and a TODO_TECHDEBT comment suggesting centralizing stake-related verification.

12. `Makefile`: Changes in this file include modifications to various targets related to stake applications and suppliers, as well as the addition of new targets for querying balance.

13. `Tiltfile`: Changes in this file include various updates to imports, comments, code formatting, and command configurations.

14. `msg_server_stake_supplier_test.go`: Changes in this file include comments and a TODO comment related to stake supplier simulation.

15. `helpers.go`: Changes in this file include the modification of an import statement and the addition of a `IsValidEndpointUrl` function to validate URLs.

16. `message_stake_application_test.go`: Changes in this file include the renaming of a comment and the addition of a new test case for valid service configurations.

17. `msg_server_unstake_application_test.go`: Changes in this file include renaming variables and updating assertions for improved readability.

18. `errors.go`: Changes in this file include the addition of new error variables and the updating of existing error variables with more descriptive error messages.

19. `tx_stake_application_test.go` in the `x/application/client/cli` directory: Changes in this file include updates to the error values of test cases, suggesting a change from invalid stake to invalid service configurations.

20. `msg_server_stake_supplier_test.go` in the `x/supplier/keeper` package: Changes in this file include the addition of an import statement, code for creating and verifying a new supplier, code for updating the stake and services of the supplier, and tests for failure scenarios.

21. `supplier_test.go`: Changes in this file include new import statements, modifications to existing functions, and the addition of a new test function.

Please review these changes thoroughly and address any necessary actions or updates.

## Issue

Part of:
- #7

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_develop_and_test`
- [x] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: proto/pocket/supplier/tx.proto [hunk 1]
```diff
@@ -6,6 +6,7 @@ import "cosmos_proto/cosmos.proto";
 import "cosmos/base/v1beta1/coin.proto";
 import "cosmos/msg/v1/msg.proto";
```
Comment (h5law) [inline proto/pocket/supplier/tx.proto:7]:
```suggestion
import "cosmos/msg/v1/msg.proto";

```

File: testutil/network/network.go [hunk 2]
```diff
@@ -117,6 +115,7 @@ func DefaultApplicationModuleGenesisState(t *testing.T, n int) *apptypes.Genesis
 				},
 			},
 		}
+		// NB: Might be useful for development but commented to cause bugs if empty fields cause business logic issues
```
Comment (h5law) [inline testutil/network/network.go:118]: I am unsure on what this means? Maybe this is a good time for us to enforce the usage of the `(gogoproto.nullable) = false` proto option?

File: x/application/types/genesis.go [hunk 3]
```diff
@@ -56,9 +56,8 @@ func (gs GenesisState) Validate() error {
 		}
 
 		// Valid the application service configs
-		// Validate the application service configs
-		if reason, ok := servicehelpers.AreValidAppServiceConfigs(app.ServiceConfigs); !ok {
-			return sdkerrors.Wrapf(ErrAppInvalidStake, reason)
+		if err := servicehelpers.AreValidAppServiceConfigs(app.ServiceConfigs); err != nil {
```
Comment (h5law) [inline x/application/types/genesis.go:59]:
```suggestion
		if err := servicehelpers.ValidateAppServiceConfigs(app.ServiceConfigs); err != nil {
```

File: x/shared/helpers/service_configs.go [hunk 4]
```diff
@@ -6,28 +6,93 @@ import (
 	sharedtypes "pocket/x/shared/types"
 )
 
-// AreValidAppServiceConfigs returns an error if the provided service configs are invalid
-// by wrapping the provided around with additional details
-func AreValidAppServiceConfigs(services []*sharedtypes.ApplicationServiceConfig) (string, bool) {
+// AreValidAppServiceConfigs returns an error if any of the application service configs are invalid
+func AreValidAppServiceConfigs(services []*sharedtypes.ApplicationServiceConfig) error {
```
Comment (h5law) [inline x/shared/helpers/service_configs.go:10]:
```suggestion
func ValidateAppServiceConfigs(services []*sharedtypes.ApplicationServiceConfig) error {
```

File: x/shared/helpers/service_configs.go [hunk 5]
```diff
@@ -6,28 +6,93 @@ import (
...
-			return fmt.Sprintf("invalid serviceId.Name: %v", serviceConfig), false
+			return fmt.Errorf("invalid serviceId.Name: %v", serviceConfig)
 		}
 	}
-	return "", true
+	return nil
+}
+
+// AreValidSupplierServiceConfigs returns an error if any of the supplier service configs are invalid
+func AreValidSupplierServiceConfigs(services []*sharedtypes.SupplierServiceConfig) error {
```
Comment (h5law) [inline x/shared/helpers/service_configs.go:35]: See above comment

File: x/supplier/client/cli/tx_stake_supplier.go [hunk 6]
```diff
@@ -17,15 +20,19 @@ var _ = strconv.Itoa(0)
 func CmdStakeSupplier() *cobra.Command {
 	// fromAddress & signature is retrieved via `flags.FlagFrom` in the `clientCtx`
 	cmd := &cobra.Command{
-		Use:   "stake-supplier [amount]",
+		// TODO_HACK: For now we are only specifying the service IDs as a list of of strings separated by commas.
```
Comment (h5law) [inline x/supplier/client/cli/tx_stake_supplier.go:23]: great comment

File: x/supplier/client/cli/tx_stake_supplier.go [hunk 7]
```diff
@@ -17,15 +20,19 @@ var _ = strconv.Itoa(0)
...
+		// This needs to be expand to specify the full SupplierServiceConfig. Furthermore, providing a flag to
+		// a file where SupplierServiceConfig specifying full service configurations in the CLI by providing a flag that accepts a JSON string
+		Use:   "stake-supplier [amount] [svcId1;url1,svcId2;url2,...,svcIdN;urlN]",
 		Short: "Stake a supplier",
 		Long: `Stake an supplier with the provided parameters. This is a broadcast operation that
 will stake the tokens and associate them with the supplier specified by the 'from' address.
```
Comment (h5law) [inline x/supplier/client/cli/tx_stake_supplier.go:29]: NIT: Until we pass in the service config file can you give a brief explainer of the comma separated format and the ; delim usage within it

File: x/supplier/client/cli/tx_stake_supplier.go [hunk 8]
```diff
@@ -54,3 +67,32 @@ $ pocketd --home=$(POCKETD_HOME) tx supplier stake-supplier 1000upokt --keyring-
...
 	return cmd
 }
+
+// TODO_BLOCKER, TODO_HACK: The supplier stake command should take an argument
+// or flag that points to a file containing all the services configurations & specifications.
+// As a quick workaround, we just need the service & url to get things working for now.
+func hackStringToServices(servicesArg string) ([]*sharedtypes.SupplierServiceConfig, error) {
+	supplierServiceConfig := make([]*sharedtypes.SupplierServiceConfig, 0)
+	serviceStrings := strings.Split(servicesArg, ",")
+	fmt.Println("OLSH", serviceStrings, servicesArg)
```
Comment (h5law) [inline x/supplier/client/cli/tx_stake_supplier.go:77]: Is this a debug line?

**Review by h5law — CHANGES_REQUESTED** [review]
Overall great PR, left a few NITs on naming mostly. But besides from those LGTM the merge is in sight :D

File: x/supplier/client/cli/tx_stake_supplier.go (hunk 8 above)
Comment (Olshansk) [reply x/supplier/client/cli/tx_stake_supplier.go:77]: Nah man, gotta leave a trail with your mark in the code ;)

File: testutil/network/network.go (hunk 2 above)
Comment (Olshansk) [reply testutil/network/network.go:118]: Replacing with `// TODO_CONSIDERATION: Evaluate whether we need `nullify.Fill` or if we should enforce `(gogoproto.nullable) = false` everywhere` everywhere

File: x/supplier/client/cli/tx_stake_supplier.go (hunk 7 above)
Comment (Olshansk) [reply x/supplier/client/cli/tx_stake_supplier.go:29]: Updated long usage with

```
TODO_HACK: Until proper service configuration files are supported, suppliers must specify the services as a single string
of comma separated values of the form 'service;url' where 'service' is the service ID and 'url' is the service URL.
For example, an application that stakes for 'anvil' could be matched with a supplier staking for 'anvil;http://anvil:8547'.
```

File: x/application/types/genesis.go [hunk 9]
```diff
@@ -34,7 +34,7 @@ func (gs GenesisState) Validate() error {
 		applicationIndexMap[index] = struct{}{}
 	}
 
-	// Check that the stake value for the apps is valid
+	// Check that the stake value for the apps is valid and that the delegatee pubkeys are valid
```
Comment (h5law) [inline x/application/types/genesis.go]:
```suggestion
	// Check that the stake value for the apps is valid and that the delegatee addresses are valid
```
must've missed this one

File: x/supplier/types/genesis.go [hunk 10]
```diff
@@ -36,6 +37,8 @@ func (gs GenesisState) Validate() error {
 
 	// Check that the stake value for the suppliers is valid
 	for _, supplier := range gs.SupplierList {
+		// TODO_TECHDEBT: Consider creating shared helpers across the board for stake validation,
+		// similar to how we have `AreValidAppServiceConfigs` below
```
Comment (h5law) [inline x/supplier/types/genesis.go:41]:
```suggestion
		// similar to how we have `ValidateAppServiceConfigs` below
```

File: x/supplier/types/genesis.go [hunk 11]
```diff
@@ -52,6 +55,12 @@ func (gs GenesisState) Validate() error {
 		if stake.Denom != "upokt" {
 			return sdkerrors.Wrapf(ErrSupplierInvalidStake, "invalid stake amount denom for supplier %v", supplier.Stake)
 		}
+
+		// Valid the application service configs
```
Comment (h5law) [inline x/supplier/types/genesis.go:59]:
```suggestion
```

File: x/supplier/types/genesis.go [hunk 12]
```diff
@@ -52,6 +55,12 @@ func (gs GenesisState) Validate() error {
...
+		// Validate the application service configs
```
Comment (h5law) [inline x/supplier/types/genesis.go:60]:
```suggestion
		// Validate the supplier service configs
```

File: x/shared/helpers/service_configs.go [hunk 13]
```diff
@@ -6,28 +6,93 @@ import (
...
-			return fmt.Sprintf("invalid serviceId.Name: %v", serviceConfig), false
+			return fmt.Errorf("invalid serviceId.Name: %v", serviceConfig)
 		}
 	}
-	return "", true
+	return nil
+}
+
+// ValidateSupplierService returns an error if any of the supplier service configs are invalid
+func ValidateSupplierService(services []*sharedtypes.SupplierServiceConfig) error {
```
Comment (h5law) [inline x/shared/helpers/service_configs.go:35]:
```suggestion
func ValidateSupplierServiceConfigs(services []*sharedtypes.SupplierServiceConfig) error {
```
NIT: Just to make it the same as the other not major tho

File: x/supplier/types/message_stake_supplier.go [hunk 14]
```diff
@@ -67,5 +72,10 @@ func (msg *MsgStakeSupplier) ValidateBasic() error {
 		return sdkerrors.Wrapf(ErrSupplierInvalidStake, "invalid stake amount denom for supplier %v", msg.Stake)
 	}
 
+	// Validate the application service configs
```
Comment (h5law) [inline x/supplier/types/message_stake_supplier.go:75]:
```suggestion
	// Validate the supplier service configs
```

**Review by h5law — APPROVED** [review]
LGTM left a few comments on some naming and comments, but once they are in lets :shipit:

**Pull request merged**